octree/include/pcl/octree/octree_base.h

```diff
@@ -102,6 +103,11 @@ namespace pcl
         BreadthFirstIterator breadth_begin(unsigned int max_depth_arg = 0) {return BreadthFirstIterator(this, max_depth_arg);};
         const BreadthFirstIterator breadth_end() {return BreadthFirstIterator();};
 
+        // Octree breadth iterators at a given depth
+        typedef OctreeBreadthFirstAtDepthIterator<OctreeT> BreadthFirstAtDepthIterator;
+        typedef const OctreeBreadthFirstAtDepthIterator<OctreeT> ConstBreadthFirstAtDepthIterator;
+        BreadthFirstAtDepthIterator breadth_at_depth_begin(unsigned int depth_arg = 0) {return BreadthFirstAtDepthIterator(this, depth_arg);};
+        const BreadthFirstAtDepthIterator breadth_at_depth_end() {return BreadthFirstAtDepthIterator();};
```

Comment:
Same as previous comment

Reply:
I've been trying to figure out how exactly this iterator `end()` is working and it seems a little different than what one would normally expect. So semantically, `end()` should point to the element right after the last one, but it seems that in this case we set the iterator to an "unitialized state" with no octree attached to it. Is my assessment correct?

Reply:
When I run `pcl_octree_viewer` with a debugger to understand (again) what's happen, on the call to the method `breadth_at_depth_end()`, I get the following call stack (I delete the template paramters):
```
#0  pcl::octree::OctreeIteratorBase::OctreeIteratorBase (this=0x7fffffffddd0, max_depth_arg=0) at /home/frozar/soft/pcl/octree/include/pcl/octree/octree_iterator.h:91
#1  pcl::octree::OctreeBreadthFirstIterator::OctreeBreadthFirstIterator (this=0x7fffffffddd0, max_depth_arg=0) at /home/frozar/soft/pcl/octree/include/pcl/octree/impl/octree_iterator.hpp:181
#2  pcl::octree::OctreeBreadthFirstAtDepthIterator::OctreeBreadthFirstAtDepthIterator (this=0x7fffffffddd0, depth_arg=0) at /home/frozar/soft/pcl/octree/include/pcl/octree/impl/octree_iterator.hpp:288
#3 pcl::octree::OctreeBase::breadth_at_depth_end (this=0x7fffffffe268) at /home/frozar/soft/pcl/octree/include/pcl/octree/octree_base.h:110
```

So calling `breadth_at_depth_end()` finally hits the constructor `OctreeIteratorBase (max_depth_arg=0)`. So the iterator is in a "sentinal state" if I can say. In my opinion it's not an "unitialized state", but still, this doesn't mean it's the good way to do it.

Actually, for an iterator the really matter is to check that after enough call to `operator++`, will it be equal to the "end iterator"? In this case, the `OctreeBreadthFirstAtDepthIterator` iterator implementation relies almost completly on the `OctreeBreadthFirstIterator` iterator. The implementation I propose mimic the behavior of `OctreeBreadthFirstIterator` iterator for this point.

What do you think about this behavior?